### ISSUE TITLE
GetWebLoginClientContext gives "out-of-date browser" problem

### ISSUE THREAD

**Issue by jkiller295** [pnp/pnpframework#832]
Following this [issue](https://github.com/pnp/pnpframework/issues/94), I have copied the code for `GetWebLoginClientContext` and implemented it in my project. However, I get this out-of-date browser error because the code uses `System.Windows.Forms.WebBrowser`  which uses the old built-in browser. 
![Capture](https://user-images.githubusercontent.com/4377795/216485716-c38ec57a-2afa-4ac1-b8d0-0b8924b51267.JPG)

What are the alternative ways of implementing this? My client has MFA turned on so the application needs to have this Web Login to go through the MFA steps.

**Comment by henrik0805** [comment]
If you e.g. try to authenticate against SharePoint Online (SPO) the message comes as EI11 is used as the browser-controller. I have rewritten the BrowserHelpder.cs to use WebView2 (i.e. EDGE) instead if interested...

**Comment by mgordic** [comment]
> If you e.g. try to authenticate against SharePoint Online (SPO) the message comes as EI11 is used as the browser-controller. I have rewritten the BrowserHelpder.cs to use WebView2 (i.e. EDGE) instead if interested...

Hi @henrik0805 , any chance to share it here?

**Comment by henrik0805** [comment]
Sorry - Was on leave, so not checking my emails.

Happy to share - You need to add the Webview2 controller as reference (e.g. via NUGET)
Also, in the setup-project, you may need to add WebView2Loader.dll (just in case the user does not have it) - It's contained in the WebView2 SDK (https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/static)
Note - My app is .NET 4.72 (not CORE, but see notes on CORE in the code - Not tested though)

Also, I am not programmer by education, so please provide feedback on code.

This us my adjusted function in BrowserHelper.js / GetWebLoginClientContext

` 


      internal static ClientContext GetWebLoginClientContext(string siteUrl, bool clearCookies, bool scriptErrorsSuppressed = true, Uri loginRequestUri = null, AzureEnvironment azureEnvironment = AzureEnvironment.Production)
        {


            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                {
                    var authCookiesContainer = new CookieContainer();
                    var siteUri = new Uri(siteUrl);
                    var cookieUrl = $"{siteUri.Scheme}://{siteUri.Host}";
                    var thread = new Thread(async () =>
                    {
                        if (clearCookies)
                        {
                            CookieReader.SetCookie(cookieUrl, "FedAuth", "ignore;expires=Mon, 01 Jan 0001 00:00:00 GMT");
                            CookieReader.SetCookie(cookieUrl, "rtFa", "ignore;expires=Mon, 01 Jan 0001 00:00:00 GMT");
                            CookieReader.SetCookie(cookieUrl, "EdgeAccessCookie", "ignore;expires=Mon, 01 Jan 0001 00:00:00 GMT");
                        }
                        System.Windows.Forms.Form form = new System.Windows.Forms.Form();

                        WebView2 webView2Control = new WebView2();

                       

                        // Customize WebView2 properties
                        webView2Control.Dock = System.Windows.Forms.DockStyle.Fill;

                        form.SuspendLayout();
                        form.Icon = null;
                        form.Width = 1024;
                        form.Height = 768;
                        form.MinimizeBox = false;
                        form.MaximizeBox = false;
                        form.Text = $"Log in to {siteUrl}";

                        // Add WebView2 control to the form
                        form.Controls.Add(webView2Control);
                        InitializeBrowser(webView2Control, siteUri.ToString());

                        webView2Control.NavigationCompleted += async (sender, args) =>
                        {
                            if ((loginRequestUri ?? siteUri).Host.Equals(webView2Control.Source.Host))
                            {
                                //var cookieString = CookieReader.GetCookie(siteUrl).Replace("; ", ",").Replace(";", ",");


                                CoreWebView2CookieManager cookieManager = webView2Control.CoreWebView2.CookieManager;
                                IList<CoreWebView2Cookie> authCookies = await cookieManager.GetCookiesAsync(siteUrl);

                                if (authCookies != null)
                                {
                                    // Set the authentication cookies both on the SharePoint Online Admin as well as on the SharePoint Online domains to allow for APIs on both domains to be used
                                    //var authCookiesString = string.Join(",", authCookies);
                                    //authCookiesContainer.SetCookies(siteUri, authCookiesString);
                                    var extension = Framework.AuthenticationManager.GetSharePointDomainSuffix(azureEnvironment);
                                    var cookieCollection = new CookieCollection();
                                    foreach (var cookie in authCookies)
                                    {
                                        var cookieName = cookie.Name;
                                        var cookieValue = cookie.Value ;
                                        cookieCollection.Add(new Cookie(cookieName, cookieValue));
                                    }
                                    authCookiesContainer.Add(new Uri(cookieUrl), cookieCollection);
                                    var adminSiteUri = new Uri(siteUri.Scheme + "://" + siteUri.Authority.Replace($".sharepoint.{extension}", $"-admin.sharepoint.{extension}"));
                                    authCookiesContainer.Add(adminSiteUri, cookieCollection);
                                    form.Close();
                                }
                            }
                        };
                        form.Focus();
                        form.ResumeLayout(false);

                        form.ShowDialog();
                        webView2Control.Dispose();
                    });
    
                    thread.SetApartmentState(ApartmentState.STA);
                    thread.Start();
                    thread.Join();
                    if (authCookiesContainer.Count > 0)
                    {
                        var ctx = new ClientContext(siteUrl);
                        ctx.DisableReturnValueCache = true;

                        // We only have to add a request digest when running in dotnet core
                        //var requestDigestInfo = GetRequestDigestAsync(siteUrl, authCookiesContainer).GetAwaiter().GetResult();
                        //requestDigestInfos.AddOrUpdate(siteUrl, requestDigestInfo, (key, oldValue) => requestDigestInfo);
                        //expiresOn = requestDigestInfo.expiresOn;

                        ctx.ExecutingWebRequest += (sender, e) =>
                        {
                            e.WebRequestExecutor.WebRequest.CookieContainer = authCookiesContainer;

                        };
    
                        var settings = new PnP.Framework.Utilities.Context.ClientContextSettings();
                        settings.Type = PnP.Framework.Utilities.Context.ClientContextType.Cookie;
                        settings.AuthenticationManager = new PnP.Framework.AuthenticationManager();
                        settings.AuthenticationManager.CookieContainer = authCookiesContainer;
                        settings.SiteUrl = siteUrl;
    
                        ctx.AddContextSettings(settings);
                        return ctx;
                    }
                }
            return null;
        }
`

**Comment by jkiller295** [comment]
@henrik0805 
There's a slight problem with using Webview2 and I don't know how to solve, after logging in to a SP site and user selects "Remember Me", if user navigates to another SP tenant, it will throw 403 Forbidden error.

**Comment by henrik0805** [comment]
Yes - The auth cookies then needs to be disposed, for the new auth to kick in..
I reset the context variable using this:
`SP.ClientContext ContextReset = BrowserHelper.GetWebLoginClientContext("#Tenant#_layouts/15/SignOut.aspx", false);`

If you need to be able to authenticate into multiple tenants at the same time, would it help to amend the "IntitializeBrowser"?
What I mean is to manage this line:
   ```
            var userDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\SPContext";
               var userDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\SPContextNewTenant";

```

Is this what you mean or?

**Comment by jkiller295** [comment]
> Yes - The auth cookies then needs to be disposed, for the new auth to kick in.. I reset the context variable using this: `SP.ClientContext ContextReset = BrowserHelper.GetWebLoginClientContext("#Tenant#_layouts/15/SignOut.aspx", false);`
> 
> If you need to be able to authenticate into multiple tenants at the same time, would it help to amend the "IntitializeBrowser"? What I mean is this line needs to handled multiple tennants..:
> 
> ```
>          var userDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\SPContext";
>             var userDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\SPContextNewTenant";
> ```
> 
> Is this what you mean or?

no, I was thinking about deleting the cookies of 1 tenant and then logging in to the new one, not at the same time. I tried expiring the cookies like this 
`CookieReader.SetCookie(cookieUrl, "FedAuth", "ignore;expires=Mon, 01 Jan 0001 00:00:00 GMT");
                    CookieReader.SetCookie(cookieUrl, "rtFa", "ignore;expires=Mon, 01 Jan 0001 00:00:00 GMT");
                    CookieReader.SetCookie(cookieUrl, "EdgeAccessCookie", "ignore;expires=Mon, 01 Jan 0001 00:00:00 GMT");`
but it stays logged in and returns 403 error when changing the tenant.

**Comment by henrik0805** [comment]
OK - Did you get this to work (i.e. log the user out)
`SP.ClientContext ContextReset = BrowserHelper.GetWebLoginClientContext("#Tenant#_layouts/15/SignOut.aspx", false);
`
Where Tenant is e.g. https://xxx.sharepoint.com/teams/yyyyy
